Ensure multiples schedules of the same job with different chains of dependent jobs are properly processed

nulltoken opened the pull request:
## Pull request description
This partly fixes #108 with regards to not properly honoring different dependent schedules.

However, this doesn't deal with the other broader concerns (for instance, how to selectively run an instant job targeting a specified configured schedule among other ones).

### PR meta checklist
- [x] Pull request is targeted at `main` branch for code   
- [x] Pull request is linked to all related issues, if any.

### Code PR specific checklist
- [ ] My code follows the code style of this project and AspNetCore coding guidelines.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [x] I have updated the appropriate sub section in the _CHANGELOG.md_.
- [x] I have added, updated or removed tests to according to my changes.
  - [x] All tests passed.

nulltoken commented on src/NCronJob/Registry/JobDefinitionEqualityComparer.cs:
```diff
@@ -0,0 +1,25 @@
+                                     && x.Type == y.Type && x.Type != typeof(DynamicJobFactory)
+                                     && x.Parameter == y.Parameter
+                                     && x.CronExpression == y.CronExpression
+                                     //&& x.TimeZone == y.TimeZone
```
@linkdotnet That removal partly played some role into being able to identify the jobDefinition in the HashSet. But I feel there are a lot of corner cases where this could blow back.

Thoughts?

![wthgcgFiaoLsHkCxVn5oS8](https://github.com/user-attachments/assets/346e475e-2170-478c-8544-af03c99aca52)

nulltoken replied:
Scratch that. I went a different way and maybe understood one of the underlying problem I was facing here.

nulltoken commented on src/NCronJob/Registry/JobDefinitionEqualityComparer.cs:
```diff
@@ -0,0 +1,25 @@
+        obj.Type,
+        obj.Parameter,
+        obj.CronExpression,
+        //obj.TimeZone,
```
Here as well

nulltoken replied:
This change has been moved back to JobRegistry

nulltoken commented on src/NCronJob/Registry/JobRegistry.cs:
```diff
@@ -19,7 +20,7 @@ internal sealed class JobRegistry
 
     public JobDefinition? FindJobDefinition(Type type)
-        => allJobs.FirstOrDefault(j => j.Type == type);
+        => allJobs.FirstOrDefault(j => j.Type == type); // TODO: We may have more than one job definition per type
```
Dragons, dragons, dragons...

nulltoken commented on src/NCronJob/Registry/JobRegistry.cs:
```diff
@@ -168,4 +172,37 @@ public int GetHashCode(JobDefinition obj) => HashCode.Combine(
+
+    private sealed class DependentJobDefinitionEqualityComparer : IEqualityComparer<JobDefinition>
+    {
+        // TODO: Maybe is the code conflating two different concepts.
```
@linkdotnet Thoughts?

linkdotnet replied:
DependentJobDefinition - that is a good question. And I can't answer the question until now.

We need a definition of a dependent job and when it gets triggered. 
Will have a look in more depth later today

linkdotnet replied:
We can invest some time into a new type `DependentJobDefinition`. It feels like it might be useful.
But not in this PR :D

linkdotnet commented on src/NCronJob/Configuration/Builder/DependentJobRegistryEntry.cs:
```diff
@@ -2,7 +2,6 @@ namespace NCronJob;
 internal sealed class DependentJobRegistryEntry
 {
-    public required Type PrincipalType { get; init; }
     public List<JobDefinition> RunWhenSuccess { get; init; } = [];
```
Could this life directly on the parent JobDefinition?

I don't see much of a need for `DepdentJobRegistryEntry` (and friends) anymore.

The question is how we can model something like:
```csharp
AddNcronJob(s => {
  s.AddJob<PrincipalJob>(...).ExecuteWhen(r => r.RunJob<DependentJob>());
  s.AddJob<DependentJob>(r => r.WithCronExpression(...)).ExecutenWhen(r => r.RunJob<DepDepJob>());
```

So if `PrincipalJob` runs, `DependentJob` should run afterward and `DepDepJob` after that.

nulltoken replied:
> Could this life directly on the parent JobDefinition?
> 
> I don't see much of a need for `DepdentJobRegistryEntry` (and friends) anymore.

If that's ok with you, I'd rather work on this in a follow up PR as an internal refactoring.

This one is already a bit "stuffy".

linkdotnet replied:
Agreed.
> Does the test do what you'd expect?
Yes

linkdotnet commented:
One test seems to be stuck:
<img width="1507" alt="image" src="https://github.com/user-attachments/assets/b4e8dff5-6532-492a-979e-6731c642c6f3">

linkdotnet commented:
Locally I can't reproduce the hanging test, but when running via Github Codespace, it also sometimes hangs. Just doesn't show the current executed test.

nulltoken commented:
> Locally I can't reproduce the hanging test, but when running via Github Codespace, it also sometimes hangs. Just doesn't show the current executed test.

On it.

nulltoken commented:
> > Locally I can't reproduce the hanging test, but when running via Github Codespace, it also sometimes hangs. Just doesn't show the current executed test.
> 
> On it.

Can't repro either.
Rebased and added some xunit configuration to list hanging tests in the future.

linkdotnet commented:
I'll try to reproduce via the Codespace. Just FYI: You can also use codespaces with 120 CPU hours for free.

linkdotnet commented:
It happes on the Codespace **only** on `net9.0`:
<img width="1165" alt="image" src="https://github.com/user-attachments/assets/979483b5-fc44-453f-82fd-f99c49943f72">

**CanBuildAChainOfDependentJobsThatRunAfterOneJob**

nulltoken commented:
> It happes on the Codespace **only** on `net9.0`: <img alt="image" width="1165" src="https://private-user-
>
> **CanBuildAChainOfDependentJobsThatRunAfterOneJob**

Thanks for the identification. I'll take a look at it tomorrow.

nulltoken commented:
> > It happes on the Codespace **only** on `net9.0`: <img alt="image" width="1165" src="https://private-user-
> > **CanBuildAChainOfDependentJobsThatRunAfterOneJob**
> 
> Thanks for the identification. I'll take a look at it tomorrow.

Actually, couldn't resist. So I took a quick peek.

The hanging problem seems unrelated to this PR as it can be reproduced on Codespaces against the `main` branch.

linkdotnet commented:
> > > It happes on the Codespace **only** on `net9.0`: <img alt="image" width="1165" src="https://private-user-
> > > **CanBuildAChainOfDependentJobsThatRunAfterOneJob**
> > 
> > 
> > Thanks for the identification. I'll take a look at it tomorrow.
> 
> Actually, couldn't resist. So I took a quick peek.
> 
> The hanging problem seems unrelated to this PR as it can be reproduced on Codespaces against the `main` branch.

Interesting! Thanks for the heads up! Well - probably it is fine if we tackle it on `main` then.

nulltoken commented:
> Interesting! Thanks for the heads up! Well - probably it is fine if we tackle it on `main` then.

Tracked through #125

linkdotnet commented on src/NCronJob/Registry/JobRegistry.cs:
```diff
@@ -19,7 +21,7 @@ internal sealed class JobRegistry
     public JobDefinition GetJobDefinition<T>() => allJobs.First(j => j.Type == typeof(T));
 
     public JobDefinition? FindJobDefinition(Type type)
```
I think we can remove the "TODO" itself. In its current state we only have a jobdefinition.

nulltoken replied:
In that context, how about replacing `FirstOrDefault()` by `SingleOrDefault()`?

This would enforce the expected intent of the design or would allow to not swallow potential issues.

linkdotnet replied:
I am a big fan of "fail fast". Do we have an option for `v3` to throw an exception when we add a duplicate? 
In v4 `UseNCronJob` could be our savior - but not sure what can be done for v3.

If there isn't anything, maybe a custom exception might be more appropriate. Not sure what a user can do if he receives the `InvalidOperationException` with "Sequence contains more than one element" message.

nulltoken replied:
I've eventually only removed the TODO.

The code actually doesn't allow duplicates. In that context, Using `FirstOrDefault()` or `SingleOrDefault()` is similar.

However, the way the code deals with duplicates isn't uniform, from a user standpoint. A such, I've created #128 (which I think should be dealt with in the context of v4 as it will certainly generate a change the behavior).

linkdotnet commented on src/NCronJob/Execution/JobExecutor.cs:
```diff
@@ -154,8 +154,8 @@ public void InformDependentJobs(JobExecutionContext context, bool success)
         var dependencies = success
-            ? jobRegistry.GetDependentSuccessJobTypes(jobRun.JobDefinition.Type)
-            : jobRegistry.GetDependentFaultedJobTypes(jobRun.JobDefinition.Type);
+            ? jobRegistry.GetDependentSuccessJobTypes(jobRun.JobDefinition)
```
Would it make sense to remove the jobRegistry's `GetDependentSuccessJobTypes` and hold this information directly on the jobdefinition or would it be to convoluted?

nulltoken replied:
I don't know yet. I'll work on this and `DepedentJobRegistryEntry` in a follow up PR.

linkdotnet commented on src/NCronJob/Registry/JobRegistry.cs:
```diff
@@ -167,4 +171,37 @@ public int GetHashCode(JobDefinition obj) => HashCode.Combine(
+            (x is null && y is null) || (x is not null && y is not null
+                                         && x.Type == y.Type && x.Type != typeof(DynamicJobFactory)
+                                         && x.Parameter == y.Parameter
+                                         //&& x.CronExpression == y.CronExpression
```
If not needed, I would remove the commented out code.

nulltoken replied:
No problem. I left it commented for you to notice the change. I'll drop it with the next push.

nulltoken replied:
Fixed

linkdotnet commented on tests/NCronJob.Tests/NCronJob.Tests.csproj:
```diff
@@ -33,12 +33,15 @@
+    <ProjectReference Include="..\..\src\NCronJob\NCronJob.csproj" />
+  </ItemGroup>
 
+  <ItemGroup>
```
Can be removed and should be tackled with #125

nulltoken replied:
Fixed

linkdotnet commented on tests/NCronJob.Tests/xunit.runner.json:
```diff
@@ -0,0 +1,5 @@
+{
```
The same here.

nulltoken replied:
Fixed

linkdotnet commented on src/NCronJob/Configuration/Builder/NCronJobOptionBuilder.cs:
```diff
@@ -190,13 +195,16 @@ internal class StartupStage<TJob> : IStartupStage<TJob> where TJob : class, IJob
     private readonly ConcurrencySettings settings;
     private readonly JobRegistry jobRegistry;
     private readonly JobOptionBuilder jobOptionBuilder;
+    private readonly ICollection<JobDefinition> jobDefinitions;
```
nitpick: I would like to keep it a `IReadOnlyCollection` so that the API is telling that this collection isn't going to change somewhere.

nulltoken replied:
Fixed

linkdotnet commented on src/NCronJob/Configuration/Builder/NCronJobOptionBuilder.cs:
```diff
@@ -246,15 +254,18 @@ internal class NotificationStage<TJob> : INotificationStage<TJob> where TJob : c
 
     internal NotificationStage(
         IServiceCollection services,
+        ICollection<JobDefinition> jobDefinitions,
```
Some would apply here

nulltoken replied:
Fixed here as well and everywhere else.

nulltoken commented:
Changelog has been updated as well.

linkdotnet commented:
Good work! I will directly make a new release!
